packages/api-server/src/methods/normalizer.ts

```diff
@@ -0,0 +1,118 @@
+import { HexNumber, HexString } from "@ckb-lumos/base";
+import { GodwokenClient } from "@godwoken-web3/godwoken";
+import { gwConfig } from "../base";
+import { EthRegistryAddress } from "../base/address";
+import { CKB_SUDT_ID, POLY_MAX_BLOCK_GAS_LIMIT } from "./constant";
+import { TransactionCallObject } from "./types";
+import {
+  verifyEnoughBalance,
+  verifyGasLimit,
+  verifyIntrinsicGas,
+} from "./validator";
+
+export class EthNormalizer {
+  private rpc: GodwokenClient;
+
+  constructor(rpc: GodwokenClient) {
+    this.rpc = rpc;
+  }
+
+  async normalizeCallTx(
+    txCallObj: TransactionCallObject
+  ): Promise<Required<TransactionCallObject>> {
+    const value = txCallObj.value || "0x0";
+    const data = txCallObj.data || "0x";
+    const toAddress = txCallObj.to;
+    const fromAddress =
+      txCallObj.from || (await getDefaultFromAddress(this.rpc));
+
+    // we should set default price to 0 instead of minGasPrice,
+    // otherwise read operation might fail the balance check.
+    const gasPrice = txCallObj.gasPrice || "0x0";
+
+    // TODO: set default gas limit to min(maxBlockGas, userBalanceAvailableGas)
+    const maxBlockGasLimit =
+      "0x" + BigInt(POLY_MAX_BLOCK_GAS_LIMIT).toString(16);
+    const gas = txCallObj.gas || maxBlockGasLimit;
+
+    const gasLimitErr = verifyGasLimit(gas, 0);
+    if (gasLimitErr) {
+      throw gasLimitErr.padContext(this.normalizeCallTx.name);
+    }
+
+    const intrinsicGasErr = verifyIntrinsicGas(toAddress, data, gas, 0);
+    if (intrinsicGasErr) {
+      throw intrinsicGasErr.padContext(this.normalizeCallTx.name);
+    }
+
+    // check if from address have enough balance
+    // when gasPrice in ethCallObj is provided.
+    if (txCallObj.gasPrice != null) {
+      const balanceErr = await verifyEnoughBalance(
+        this.rpc,
+        fromAddress,
+        value,
+        gas,
+        gasPrice,
+        0
+      );
+      if (balanceErr) {
+        throw balanceErr.padContext(
+          `${this.normalizeCallTx.name}: from account ${fromAddress}`
+        );
+      }
+    }
+
+    return {
+      value,
+      data,
+      to: toAddress,
+      from: fromAddress,
+      gas,
+      gasPrice,
+    };
+  }
+
+  async normalizeEstimateGasTx(
+    txEstimateGasObj: Partial<TransactionCallObject>
+  ): Promise<Required<TransactionCallObject>> {
+    const to = txEstimateGasObj.to || "0x";
+    return this.normalizeCallTx({ ...txEstimateGasObj, to });
+  }
+}
+
+async function getDefaultFromAddress(rpc: GodwokenClient): Promise<HexString> {
+  const defaultFromScript = await rpc.getScript(
+    gwConfig.accounts.defaultFrom.scriptHash
+  );
+  if (defaultFromScript == null) {
+    throw new Error("default from script is null");
+  }
+  const defaultFromAddress = "0x" + defaultFromScript.args.slice(2).slice(64);
+  return defaultFromAddress;
+}
+
+export async function getMaxGasByBalance(
```

Comment:
unused function?

Reply:
the follow up pr https://github.com/nervosnetwork/godwoken-web3/pull/520 will use it